src/librustc/ty/layout.rs

```diff
@@ -1349,14 +1345,33 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
         // get included in each variant that requested them in
         // GeneratorLayout.
         debug!("prefix = {:#?}", prefix);
-        let (outer_fields, promoted_offsets) = match prefix.fields {
-            FieldPlacement::Arbitrary { mut offsets, .. } => {
-                let offsets_b = offsets.split_off(discr_index + 1);
+        let (outer_fields, promoted_offsets, promoted_memory_index) = match prefix.fields {
+            FieldPlacement::Arbitrary { mut offsets, memory_index } => {
+                let mut inverse_memory_index = invert_mapping(&memory_index);
```

Comment:
To make all of the code dealing with indexes easier to follow, it would help to have a comment that says each memory_index is field_idx -> memory ordering, while each inverse_memory_index is the list of field indices in memory order. I still get these mixed up so I imagine other people reading will, too.

Reply:
There's something about that somewhere that `invert_mapping` is referring to, but I guess I can have a comment that's summarizing everything, copied into a couple places.

Reply:
Yeah, there are some inner comments in `univariant_uninterned`. I don't consider them especially discoverable.. maybe it would be good to have an explanation on the `FieldPlacement::Arbitrary::memory_index` field itself and maybe also a summary on `invert_mapping` (since that gets called from both here and `univariant_uninterned`.)

Reply:
Another thing I'm noticing is that [we invert the index again](https://github.com/rust-lang/rust/blob/1fc9365c8f04c121265b6b46f083e142baf44620/src/librustc_target/abi/mod.rs#L747) in codegen. It is possible it would be easier to keep it in its inverted form?

Reply:
This is intentional, `index_by_increasing_offset` is relatively rarely called (only the LLVM backend needs it, and once per layout), whereas `memory_index` is used for every field access (arguably also only by the LLVM backend, but more frequently).